musicd/dbupdater.go

```diff
@@ -0,0 +1,138 @@
+/*
+ * 
+ */
+
+package main
+
+import (
+        // "database/sql"
+	"fmt"
+	"log"
+	// "os"
+	// "strings"
+	"time"
+
+	"github.com/mattn/go-sqlite3"
+	// "github.com/spf13/viper"
+
+	"github.com/DNSSEC-Provisioning/music/common"
+)
+
+func dbUpdater(conf *Config) {
+
+     log.Printf("dbUpdater: Starting DB Update Service.")
+
+	mdb := conf.Internal.MusicDB
+
+	// dbupdatechan := conf.Internal.DBUpdater
+	dbupdateC := make(chan music.DBUpdate, 5)
+	conf.Internal.DBUpdateC = dbupdateC
+	mdb.UpdateC = dbupdateC
+
+//	var tx *sql.Tx
+//	localtx, tx, err := mdb.StartTransaction(tx)
+//	if err != nil {
+//		log.Printf("SetStopReason: Error from mdb.StartTransaction(): %v\n", err)
+//		return err, "fail"
+//	}
+//	defer mdb.CloseTransaction(localtx, tx, err)
+
+	const ZSMsql = "INSERT OR REPLACE INTO metadata (zone, key, time, value) VALUES (?, ?, datetime('now'), ?)"
+
+	mstmt, err := mdb.Prepare(ZSMsql)
+	if err != nil {
+		fmt.Printf("ZoneSetMeta: Error from db.Prepare(%s) 1: %v\n", ZSMsql, err)
+	}
+
+//	_, err = mstmt.Exec(z.Name, key, value)
+//	if CheckSQLError("ZoneSetMeta", "INSERT OR REPLACE INTO", err, false) {
+//		return err, ""
+//	}
+
+//	return nil, fmt.Sprintf("Zone %s metadata '%s' updated to be '%s'",
+//		z.Name, key, value)
+
+	const DSsql = "UPDATE zones SET fsmstatus='blocked' WHERE name=?"
+	blockstmt, err1 := mdb.Prepare(DSsql)
+	if err1 != nil {
+		log.Fatalf("DocumentStop: Error from mdb.Prepare(%s): %v", DSsql, err)
+	}
+
+//	_, err1 = stmt.Exec(z.Name)
+//	if err1 != nil {
+//		log.Fatalf("DocumentStop: Error from mdb.Exec(%s): %v", DSsql, err)
+//	}
+//	log.Printf("%s: %s\n", z.Name, value)
+
+	ticker := time.NewTicker(2 * time.Second)
+
+	queue := []music.DBUpdate{}
+	var update music.DBUpdate
+
+	RunDBQueue := func() {
+		for {
+			if len(queue) == 0 {
+				// log.Printf("RunDBQueue: DBQueue is empty")
+				break
+			}
+			u := queue[0]
+			t := u.Type
+
+			tx, err := mdb.Begin()
+			if err != nil {
+				log.Printf("RunDBQueue: Error from mdb.Begin(): %v", err)
+			}
+
+			switch t {
+			case "STOPREASON":
+				_, err := mstmt.Exec(u.Zone, u.Key, u.Value)
```

Comment:
I don't think this works as you intended it. The statement is not doing it's thing in the transaction since you created it using the mdb variable. You can read more on prepared statements in go here https://go.dev/doc/database/prepared-statements 

I would also consider not using statements or not creating them inside the functions that use them.

Reply:
Hmm? Nu blev det för komplicerat för mig. Själva vitsen med ett statement är ju att ha preppat det i förväg så att det kan återanvändas gång på gång (förutom säkerhetsaspekter som att hjälpa till att skydda mot SQL Injection, etc, men det är inte problemet just här).

Så jag förstår inte vad som är fel i göra på detta sätt.

Reply:
Tyckte att jag svarade på detta nyss, men det försvann. Jag förstår helt enkelt inte vad du tycker är fel. Att använda ett prepared statement är ju exakt för (bl a) ändamålet "compile once, use repeatedly" vilket är vad jag gör här. Så du har säkert rätt, men jag är för trög för att begripa vad som är fel.

Reply:
Det är fortfarande någonting som jag inte fattar här. I koden står det i princip:
`tx := mdb.Begin()
...
switch t {
case "FOO": _, err := stmt.Exec()
...`

På vilket sätt är inte stmt.Exec() inne i transaktionen?

Reply:
Variabeln tx är din transaktion, om du vill exekvera något inom din transaktion måste det göras med den variablen. 

Till exempel 
tx, err := mdb.Begin() <-- skapa transaktionen
mstmt, err := tx.Prepare(ZSMsql)   <-- skapa statementet inne i transaktionen
_, err := mstmt.Exec(u.Zone, u.Key, u.Value)  <-- exekvera statementet inne i transaktionen
tx.Commit() <-- här genomförs statementet genom att du commitar transaktionen

Istället för 
mstmt, err := mdb.Prepare(ZSMsql)   <-- skapa statementet utanför en transaktion
tx, err := mdb.Begin() <-- en transaktion skapas som aldrig används
_, err := mstmt.Exec(u.Zone, u.Key, u.Value)  <-- exekvera statementet utanför transaktionen
tx.Commit() <-- här genomförs ingenting för att du inte har exekverat något i transaktionen

Reply:
Statements bör också stängas när de inte används mer, så en defer stmt.Close() bör finnas efter varje Prepare()

Reply:
Isåfall är hela den omstuvning som vi just gjort felaktig. Men den funkar bra. Nu förstår jag vad det är du försöker förklara, jag beklagar att jag är så trög. Men jag undrar om orsaken till att det faktiskt fungerar är just detta med att vi bara har exakt ett databaskoppel. Så att det görs Prepare() direkt på kopplet leder till ett statement som sedan även kan användas inne i transaktionen?

Reply:
Vad sedan gäller detta med att alltid skapa statements inne i en transaktion så innebär det ju att det i princip inte går att använda statements som "speed hacks" (compile once, use over and over) annat än för trivialfallet "här har jag en färdig bunt med saker som jag skall göra samma sak med allihop just nu".

Att (som i detta fall) veta att vi har en stadig ström av db updates av en viss typ så vi accepterar resursåtgången av att ha ett färdigt statement som är redo varje gång det är dags... funkar inte. Och det stämmer ju även med din kommentar om att du tycker att vi kanske inte borde bry oss om att använda statements.

Reply:
Å andra sidan är statements ju även en mekanism för att skydda sig mot SQL Injection så det finns ju för all del även andra skäl...

Stort tack för alla dina supersnabba kommentarer. Jag tror att vi kanske borde ha ett möte bara för att diskutera SQL och hur man klappar sin databas. Jag är uppenbart rätt suddig om hur det egentligen fungerar.